### PR TITLE
Improve test for 'symbol.exports'

### LINKED ISSUE #25484: TSServer: Cannot read property 'forEach' of undefined
**TypeScript Version:**  3.0.0-dev.20180706

**Search Terms:** 

**Code**

```ts
TypeError: Cannot read property 'forEach' of undefined
    at visitSymbol (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29086:35)
    at visitSignature (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29040:21)
    at visitObjectType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29060:21)
    at visitType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28992:25)
    at visitSymbol (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29084:17)
    at visitObjectType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29068:21)
    at visitInterfaceType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29046:17)
    at visitType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28989:25)
    at visitTypeReference (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29009:17)
    at visitType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28983:25)
    at visitSignature (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29043:17)
    at visitObjectType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29060:21)
    at visitInterfaceType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29046:17)
    at visitType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28989:25)
    at visitSymbol (/usr/local/lib/node_modules/typescript/lib/tsserver.js:29084:17)
    at visitType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28975:34)
    at Object.walkType (/usr/local/lib/node_modules/typescript/lib/tsserver.js:28946:25)
    at recordTypeParameterUsages (/usr/local/lib/node_modules/typescript/lib/tsserver.js:108636:53)
    at collectUsages (/usr/local/lib/node_modules/typescript/lib/tsserver.js:108648:25)
    at visitNode (/usr/local/lib/node_modules/typescript/lib/tsserver.js:15101:24)
    at Object.forEachChild (/usr/local/lib/node_modules/typescript/lib/tsserver.js:15368:21)
    at collectUsages (/usr/local/lib/node_modules/typescript/lib/tsserver.js:108678:28)
    at visitNodes (/usr/local/lib/node_modules/typescript/lib/tsserver.js:15110:30)
    at Object.forEachChild (/usr/local/lib/node_modules/typescript/lib/tsserver.js:15334:24)
    at collectUsages (/usr/local/lib/node_modules/typescript/lib/tsserver.js:108678:28)
    at collectReadsAndWrites (/usr/local/lib/node_modules/typescript/lib/tsserver.js:108542:17)
    at getPossibleExtractionsWorker (/usr/local/lib/node_modules/typescript/lib/tsserver.js:107917:38)
    at getPossibleExtractions (/usr/local/lib/node_modules/typescript/lib/tsserver.js:107872:26)
    at Object.getAvailableActions (/usr/local/lib/node_modules/typescript/lib/tsserver.js:107383:35)
    at /usr/local/lib/node_modules/typescript/lib/tsserver.js:103417:128
    at getIterator (/usr/local/lib/node_modules/typescript/lib/tsserver.js:534:23)
    at Object.next (/usr/local/lib/node_modules/typescript/lib/tsserver.js:529:35)
    at Object.arrayFrom (/usr/local/lib/node_modules/typescript/lib/tsserver.js:1128:32)
    at Object.getApplicableRefactors (/usr/local/lib/node_modules/typescript/lib/tsserver.js:103416:23)
    at Object.getApplicableRefactors (/usr/local/lib/node_modules/typescript/lib/tsserver.js:111461:32)
    at IOSession.Session.getApplicableRefactors (/usr/local/lib/node_modules/typescript/lib/tsserver.js:118719:53)
    at Session.handlers.ts.createMapFromTemplate._a.(anonymous function) (/usr/local/lib/node_modules/typescript/lib/tsserver.js:117667:61)
    at /usr/local/lib/node_modules/typescript/lib/tsserver.js:118945:88
    at IOSession.Session.executeWithRequestId (/usr/local/lib/node_modules/typescript/lib/tsserver.js:118936:28)
    at IOSession.Session.executeCommand (/usr/local/lib/node_modules/typescript/lib/tsserver.js:118945:33)
    at IOSession.Session.onMessage (/usr/local/lib/node_modules/typescript/lib/tsserver.js:118965:35)
    at Interface.<anonymous> (/usr/local/lib/node_modules/typescript/lib/tsserver.js:120223:27)
```
tsserver logs id: 3556371867

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #25484

### CODE REVIEW

File: src/compiler/symbolWalker.ts [hunk 1]
```diff
@@ -171,8 +171,8 @@ namespace ts {
                 }
                 const t = getTypeOfSymbol(symbol);
                 visitType(t); // Should handle members on classes and such
-                if (symbol.flags & SymbolFlags.HasExports) {
-                    symbol.exports!.forEach(visitSymbol);
+                if (symbol.exports) {
```
Comment (mhegazy) [inline src/compiler/symbolWalker.ts]: so when does it have the flag but no exports? this is more concerning than the error really.

Comment (andy-ms) [reply src/compiler/symbolWalker.ts]: We don't have a repro for this unfortunately.

Comment (mhegazy) [reply src/compiler/symbolWalker.ts]: we only use `HasExports` in two places other than this one.. once in the binder and once
in the checker..  I wounder if we can just remove it all together, and check `exports` instead.

Comment (mhegazy) [reply src/compiler/symbolWalker.ts]: @weswigham and @sandersn thoughts?

Comment (weswigham) [reply src/compiler/symbolWalker.ts]: `HasExports` afaik is set on container symbols which _can_ have exports, while the member is present if there actually are any. I imagine the usage in the checker is likely incorrect - AFAIK it should only be used in the binder to determine how to initialize a symbol.

Comment (sandersn) [reply src/compiler/symbolWalker.ts]: It's only an alias, so even in the binder, the code might actually be clearer without it. (or with a comment instead)

**Comment by andy-ms** [comment]
@sandersn A JS baseline needed an update -- pretty sure this is correct since `exports.x = 0; exports.x = 1;` is an error, but just want to confirm.

**Comment by sandersn** [comment]
That baseline came from #22461, which has examples from graceful-fs, uglify and webpack. All of those assign exports to `exports` *and* to an alias, but none of those makes multiple assignments to the same property. So I think the baseline change is good.

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -225,11 +225,11 @@ namespace ts {
             node.symbol = symbol;
             symbol.declarations = append(symbol.declarations, node);
 
-            if (symbolFlags & SymbolFlags.HasExports && !symbol.exports) {
+            if (symbolFlags & (SymbolFlags.Class | SymbolFlags.Enum | SymbolFlags.Module | SymbolFlags.Variable) && !symbol.exports) {
```
Comment (sandersn) [inline src/compiler/binder.ts]: I like this change.

**Pull request merged**